state/store.go

```diff
@@ -367,7 +365,6 @@ func (store dbStore) PruneStates(from int64, to int64) error {
 			}
 			batch.Close()
 			batch = store.db.NewBatch()
-			defer batch.Close()
```

Comment:
Soliloquizing: I wasn't sure exactly which variable was used, the one bound with `defer` was _declared_ or the one when `defer` was _invoked_, but this is the right move since it seems that the later variable seems to be used.

```go
package main

import "fmt"

type X struct {
	value int
}

func (x *X) Foo() {
	fmt.Printf("Foo(%d)\n", x.value)
}

func main() {
	x := X{value: 0}
	defer x.Foo()
	x = X{value: 10}
	// --> Foo(10)
}
```

Reply:
Thanks for your checking. Here is the minimized code of the related `batch`:

```
func (store dbStore) PruneStates(from int64, to int64) error {
	batch := store.db.NewBatch()
	defer batch.Close()
	for h := to - 1; h >= from; h-- {
		if pruned%1000 == 0 && pruned > 0 {
			err := batch.Write()
			if err != nil {
				return err
			}
			batch.Close()
			batch = store.db.NewBatch()
			defer batch.Close()
		}
	}
	err = batch.WriteSync()
	if err != nil {
		return err
	}
	return nil
}
```

I've confirmed `defer` is already set, so I think it's unnecessary `defer` in the loop. But, Golang holds the `batch` until the end of the function. It means `batch` variable pointer exists the number=first one+to count. So, I'll revert my changes.

```
func Test(t *testing.T) {
	e := newEcho("hello")
	defer e.print()
	for i := 0; i < 10; i++ {
		e = newEcho(fmt.Sprintf("hello%d", i))
		defer e.print()
	}
}
type echo struct {
	hello string
}
func newEcho(hello string) echo {
	return echo{hello: hello}
}
func (e echo) print() {
	fmt.Printf("%p;%s\n", &e, e.hello)
}
```
```
=== RUN   Test
0x14000050560;hello9
0x14000050580;hello8
0x140000505a0;hello7
0x140000505c0;hello6
0x140000505e0;hello5
0x14000050600;hello4
0x14000050620;hello3
0x14000050640;hello2
0x14000050660;hello1
0x14000050680;hello0
0x140000506a0;hello
```


BTW, in this case, setting `defer` in the loop is not a good manner. Should it hold the `batch` variable of an array?
<img width="456" alt="screenshot 2022-12-09 17 14 07" src="https://user-images.githubusercontent.com/1164270/206655973-f77d5999-c0f7-424b-9976-405a22fcd0ae.png">

Reply:
Isn't there a need for `defer batch.Close()` in the loop since it closes the previous `batch` in the loop and creates a new `batch`?

Reply:
Yes, it does. In my checking with the no-pointer receiver `func (x X) Foo()`, it's necessary.

Reply:
I see. It certainly seems that if I rewrite the method called out of `defer` to be a value-receiver, it seems that the last Close is indeed not called, but instead a Close is called on the instance that should have been closed first.  I can't explain why it works that way, though.

```go
func (x X) Close() {
	fmt.Printf("%p: Close(%s)\n", &x, x.greeting)
}

func main() {
	x := X{greeting: "hello"}
	defer x.Close()
	for i := 1; i < 5; i++ {
		x.Close()
		x = X{greeting: fmt.Sprintf("hello%d", i)}
	}
	fmt.Printf("--- iteration ended\n")
}
```

```
0xc00009a050: Close(hello)
0xc00009a070: Close(hello1)
0xc00009a090: Close(hello2)
0xc00009a0b0: Close(hello3)
--- iteration ended
0xc00009a0d0: Close(hello)    <-- should be Close(hello4)
```